Lib/test/test_statistics.py

```diff
@@ -2161,6 +2161,32 @@ def test_center_not_at_mean(self):
         self.assertEqual(self.func(data), 2.5)
         self.assertEqual(self.func(data, mu=0.5), 6.5)
 
+class TestSqrtHelper(unittest.TestCase):
+
+    def test_sqrt_frac(self):
+        # This helper function aspires to produce correctly rounded square roots of
+        # fractional inputs, more accurate than can be had with math.sqrt() alone.
+        # For a inputs spanning large ranges, we test that
+        # 1) the result is close to math.sqrt()
+        # 2) the result is at least as good as the two adjacent float values
+
+        randrange = random.randrange
+
+        for i in range(50_000):
+            numerator: int = randrange(10 ** randrange(40))
+            denonimator: int = randrange(10 ** randrange(40)) + 1
+            with self.subTest(numerator=numerator, denonimator=denonimator):
+                x: Fraction = Fraction(numerator, denonimator)
+
+                root: float = statistics._sqrt_frac(numerator, denonimator)
+                self.assertTrue(math.isclose(root, math.sqrt(numerator / denonimator)))
+
+                r_up: float = math.nextafter(root, math.inf)
+                self.assertLessEqual(abs(Fraction(root)**2 - x), abs(Fraction(r_up)**2 - x))
+
+                r_down: float = math.nextafter(root, -math.inf)
+                self.assertLessEqual(abs(Fraction(root)**2 - x), abs(Fraction(r_down)**2 - x))
```

Comment:
Strictly speaking, this isn't quite the right test. We want to know that `|root - √x| ≤ |r_down - √x|`, but we're testing that `|root² - x| ≤ |r_down² - x|` instead. There are cases where those two conditions differ, though the likelihood of us actually hitting one of those cases in this test may well be so small as to be negligible.

As a (contrived but easy) example, take `x = (10**16 - 1)**2`. Then the correctly rounded square root `root` of `x` is `1e16`. `r_down` will be `1e16 - 2`, but this test will fail, because `r_down**2` is closer to `x` than `r**2` is, even though `r_down` and `r` are equally close to `√x`.

Reply:
Suggested fix: check that `√x ≥ (root + r_down) / 2` instead, by checking that `4x ≥ (root + r_down)²`; both sides can be computed exactly as fractions.

Reply:
I put in a new test before reading this.  Does it fit the bill?

Reply:
Yep, looks good to me!